ICMP: Introduce ICMP type name in ICMPField

Shunpoco opened the pull request:
Currently ICMP only supports ICMP type code (0-255), but ideally it should also support ICMP type name like "Echo", or "Echo Reply".

This commit changes the type of ICMPField.Type from uint8 to intstr.IntOrString, and also updates ICMPField.PortProtocol to treat type names.

ICMPField treats both ICMP IPv4 and v6, but kubebuilder's validation can't distinguish if the given type name matches with family (IPv4 or IPv6). So this commit also introduces UnmarshalJSON method for ICMPField. That function checks if the unmarshaled ICMPField's family and type match properly.

In addition, this commit changes some tests which use ICMPField, and introduces an unit test for ICMPField's UnmarshalJSON method.

Fixes: #23000

Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

<!-- Description of change -->

```release-notep
policy: Change the ICMPField type in ICMP policies from integers only to integers or strings.
```

Shunpoco committed:
ICMP: Introduce ICMP type name in ICMPField

Currently ICMP only supports ICMP type code (0-255), but ideally it should also support ICMP type names like "Echo", or "Echo Reply".

This commit changes the type of ICMPField.Type from uint8 to intstr.IntOrString, and also updates ICMPField.PortProtocol to treat type names.

ICMPField treats both ICMP IPv4 and v6, but kubebuilder's validation can't distinguish if the given type name matches with the family (IPv4 or IPv6). So this commit also introduces UnmarshalJSON method for ICMPField. That function checks if the unmarshaled ICMPField's family and type match properly.

In addition, this commit changes some tests which use ICMPField, and introduces a unit test for ICMPField's UnmarshalJSON method.

Fixes: #23000

Signed-off-by: Shunsuke Tokunaga <tkngsnsk313320@gmail.com>

Shunpoco committed:
doc: modify/add ICMP type change

ICMP `type` field is changed to accept both integer and string.

This commit updates:
- v1.15 Upgrade Notes to add about it.
- ICMP part in layer 4 examples

Signed-off-by: Shunsuke Tokunaga <tkngsnsk313320@gmail.com>

Shunpoco committed:
doc: modify upgrade note and comment in code block

- Modify comment in upgrade note in order to focus on user-facing
CRD change
- Fix indentation in code block

Signed-off-by: Shunsuke Tokunaga <tkngsnsk313320@gmail.com>

Shunpoco commented on Documentation/operations/upgrade.rst:
```diff
@@ -386,6 +386,10 @@ Changed Metrics
   is disabled), it will fail to initialize and will log an error instead of silently
   fall back to iptables based masquerading.
+* The Go types for Network policies have changed to accept both integers and strings in
+  the ICMP protocol type. Specifically, the ICMP ``type`` field can be either
```
Thanks. I addressed to both of your comment at https://github.com/cilium/cilium/pull/30330/commits/5ca340c7ac1b68daf83e90f3f25fa8e18266d1a2.

danehans commented on Documentation/security/policy/language.rst:
```diff
@@ -754,25 +754,45 @@ which is defined as follows:
+	        // 			 ICMP Node Information Response | Inverse Neighbor Discovery Solicitation | Inverse Neighbor Discovery Advertisement |
+	        // 			 Home Agent Address Discovery Request | Home Agent Address Discovery Reply | Mobile Prefix Solicitation |
+	        // 			 Mobile Prefix Advertisement | Duplicate Address Request Code Suffix | Duplicate Address Confirmation Code Suffix |
+	        // 			 Extended Echo Request | Extended Echo Reply
```
To provide consistency with [k8s API conventions](https://github.com/zecke/Kubernetes/blob/master/docs/devel/api-conventions.md#constants), should the allowed values be CamelCase?

nathanjsweet replied:
Yeah that's fair. @Shunpoco Can you please make that change, thanks :pray: ?

Shunpoco replied:
Fixed in https://github.com/cilium/cilium/commit/7c91e9448ba2201205da66b6b5ee09754cd2f815

danehans reviewed (approved):
Approving since my comment is non-blocking.

nathanjsweet reviewed (changes requested):
Turn the string options into camel-case rather than space delimited.

Shunpoco committed:
ICMP: Use CamelCase for ICMP type messages

This commit fixes ICMP type messages to use CamelCase instead of space-separated words.
For example, Echo Reply is changed to EchoReply.

Signed-off-by: Shunsuke Tokunaga <tkngsnsk313320@gmail.com>
